describe.go

```diff
@@ -0,0 +1,204 @@
+// either refs/tags/ or refs/ respectively.
+//
+// By default it only shows annotated tags.
+type DescribeOptionsStrategy uint
+
+// Describe strategy options.
+const (
+	DescribeDefault DescribeOptionsStrategy = C.GIT_DESCRIBE_DEFAULT
+	DescribeTags    DescribeOptionsStrategy = C.GIT_DESCRIBE_TAGS
+	DescribeAll     DescribeOptionsStrategy = C.GIT_DESCRIBE_ALL
+)
+
+// Describe performs the describe operation on the commit.
+func (c Commit) Describe(describeOpts *DescribeOptions, formatOpts *DescribeFormatOptions) (string, error) {
+	var resultPtr *C.git_describe_result
+
+	var cDescribeOpts *C.git_describe_options
+	if describeOpts != nil {
+		var cpattern *C.char
+		if len(describeOpts.Pattern) > 0 {
+			cpattern = C.CString(describeOpts.Pattern)
+			defer C.free(unsafe.Pointer(cpattern))
+		}
+
+		cDescribeOpts = &C.git_describe_options{
+			version:                     C.GIT_DESCRIBE_OPTIONS_VERSION,
+			max_candidates_tags:         C.uint(describeOpts.MaxCandidatesTags),
+			describe_strategy:           C.uint(describeOpts.Strategy),
+			pattern:                     cpattern,
+			only_follow_first_parent:    cbool(describeOpts.OnlyFollowFirstParent),
+			show_commit_oid_as_fallback: cbool(describeOpts.ShowCommitOidAsFallback),
+		}
+	}
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	ecode := C.git_describe_commit(&resultPtr, c.gitObject.ptr, cDescribeOpts)
+	if ecode < 0 {
+		return "", MakeGitError(ecode)
+	}
+	defer C.git_describe_result_free(resultPtr)
+
+	return describeFormat(resultPtr, formatOpts)
+}
+
+// DescribeWorkdir describes the working tree. It means describe HEAD
+// and appends <mark> (-dirty by default) if the working tree is dirty.
+func (repo *Repository) DescribeWorkdir(describeOpts *DescribeOptions, formatOpts *DescribeFormatOptions) (string, error) {
+	var resultPtr *C.git_describe_result
+
+	var cDescribeOpts *C.git_describe_options
+	if describeOpts != nil {
+		var cpattern *C.char
+		if len(describeOpts.Pattern) > 0 {
+			cpattern = C.CString(describeOpts.Pattern)
+			defer C.free(unsafe.Pointer(cpattern))
+		}
+
+		cDescribeOpts = &C.git_describe_options{
+			version:                     C.GIT_DESCRIBE_OPTIONS_VERSION,
+			max_candidates_tags:         C.uint(describeOpts.MaxCandidatesTags),
+			describe_strategy:           C.uint(describeOpts.Strategy),
+			pattern:                     cpattern,
+			only_follow_first_parent:    cbool(describeOpts.OnlyFollowFirstParent),
+			show_commit_oid_as_fallback: cbool(describeOpts.ShowCommitOidAsFallback),
+		}
+	}
+
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	ecode := C.git_describe_workdir(&resultPtr, repo.ptr, cDescribeOpts)
+	if ecode < 0 {
+		return "", MakeGitError(ecode)
+	}
+	defer C.git_describe_result_free(resultPtr)
+
+	return describeFormat(resultPtr, formatOpts)
+}
+
+func describeFormat(resultPtr *C.git_describe_result, formatOpts *DescribeFormatOptions) (string, error) {
```

Comment:
This function should be exported and either have a `DescribeResult` receiver or take one as a parameter.